core/src/main/java/com/scalar/db/storage/jdbc/JdbcTableMetadataManager.java

```diff
@@ -105,26 +115,213 @@ public JdbcTableMetadataManager(
+      case SQL_SERVER:
+        return value ? "1" : "0";
+      default:
+        return value ? "true" : "false";
+    }
+  }
+
+  private String getDeleteStatement(String namespace, String tableName) {
+    String fullTableName = schemaPrefix.orElse("") + namespace + "." + tableName;
+    return "DELETE FROM "
+        + enclosedFullTableName(getMetadataSchema(), TABLE, rdbEngine)
+        + " WHERE "
+        + enclose(FULL_TABLE_NAME, rdbEngine)
+        + " = '"
+        + fullTableName
+        + "'";
+  }
+
+  private void insertMetadataColumn(
+      String schema,
+      String table,
+      TableMetadata metadata,
+      Connection connection,
+      int ordinalPosition,
+      String column)
+      throws SQLException {
+    KeyType keyType = null;
+    if (metadata.getPartitionKeyNames().contains(column)) {
+      keyType = KeyType.PARTITION;
+    }
+    if (metadata.getClusteringKeyNames().contains(column)) {
+      keyType = KeyType.CLUSTERING;
+    }
+
+    try {
+      String insertStatement =
+          getInsertStatement(
+              schema,
+              table,
+              column,
+              metadata.getColumnDataType(column),
+              keyType,
+              metadata.getClusteringOrder(column),
+              metadata.getSecondaryIndexNames().contains(column),
+              ordinalPosition);
+      execute(connection, insertStatement);
+    } catch (SQLException e) {
+      connection.rollback();
+      throw e;
+    }
+  }
+
+  private void createMetadataTable() {
+    String createTableQuery =
+        String.format(
+                "CREATE TABLE %s(", enclosedFullTableName(getMetadataSchema(), TABLE, rdbEngine))
+            + String.format("%s %s,", enclose(FULL_TABLE_NAME, rdbEngine), getTextType(128))
+            + String.format("%s %s,", enclose(COLUMN_NAME, rdbEngine), getTextType(128))
+            + String.format("%s %s NOT NULL,", enclose(DATA_TYPE, rdbEngine), getTextType(20))
+            + String.format("%s %s,", enclose(KEY_TYPE, rdbEngine), getTextType(20))
+            + String.format("%s %s,", enclose(CLUSTERING_ORDER, rdbEngine), getTextType(10))
+            + String.format("%s %s NOT NULL,", enclose(INDEXED, rdbEngine), getBooleanType())
+            + String.format("%s INTEGER NOT NULL,", enclose(ORDINAL_POSITION, rdbEngine))
+            + String.format(
+                "PRIMARY KEY (%s, %s))",
+                enclose(FULL_TABLE_NAME, rdbEngine), enclose(COLUMN_NAME, rdbEngine));
+
+    try (Connection connection = dataSource.getConnection()) {
+      connection.setAutoCommit(false);
+      connection.setTransactionIsolation(Connection.TRANSACTION_SERIALIZABLE);
+      try {
+        createMetadataSchema(connection);
+        execute(connection, createTableQuery);
+      } catch (SQLException e) {
+        connection.rollback();
+        throw e;
+      }
+      connection.commit();
+    } catch (SQLException e) {
+      if (e.getMessage().contains("database exists")
```

Comment:
Not sure but checking the actual messages looks implementation-specific.
Do MySQL and PostgreSQL return the same message as long as we use the same JDBC?

Reply:
Indeed, the error message is implementation specific. 

> Do MySQL and PostgreSQL return the same message as long as we use the same JDBC?

It seems it is the case but I would have to do more extensive test. 

Otherwise I think we could either : 
- Option 1 : Rely on [SqlException.getErrorCode()](https://docs.oracle.com/javase/jp/8/docs/api/java/sql/SQLException.html#getErrorCode--) which returns a vendor specific error code. I checked will Mysql, Postgresql, SqlServer and Oracle. Each of them throw a custom error when trying to create a duplicate schema/table. I suppose such error code are stable through the different driver versions.
- Option 2 : Mysql and Postgresql supports the `IF EXISTS` keyword that we can use when creating schema and table. In that case no exception would be thrown for the duplication. 
But Oracle and Sql server do not have `IF EXISTS` keyword so we can rely on checking error code for only these two. If error code checking is not satisfying, we can run some select query towards system database to verify if such user or table already exist, which seems a bit hacky. 

What do you think ?

Reply:
If the exception object has a proper message even though they are vendor-specific or different, I think we can always throw an exception as you do now instead of doing special care depending on the messages.
(So, the if block can be simply removed)

Reply:
Maybe there was a misunderstanding but remove the exception suppression will break the current design.
 Each time the `TableMetadataManager.addTableMetadata` method is called, it tries to create the `scalardb` schema and `scalardb.metadata` table and then add the given sample_schema.sample_table table metadata. If we remove the exception suppression block in case the `scalardb.metadata` has already been created, we won't be able to run `TableMetadataManager.addTableMetadata` more than once.

Reply:
Sorry, I'm saying this without fully understanding the goal. 
OK, if the logic is required, I think we should implement the logic for each supported database like MySQL, PostgreSQL, Oracle, and SQLServer since JDBC doesn't abstract the if-exists logic.
I think the current JDBC adapter implements PutIfNotExists so you can do it in the same way at least logically.

Reply:
Ah, OK, it's for creating schema. What I was saying in the last sentence is about inserting with IF NOT EXISTS, so it's a different thing. Please ignore the last sentecne.

Reply:
Alright. 👍 
To be sure I understood correctly, I will update as follows : 
- For **Mysql** & **Postgresql**: use `IF EXISTS` statement in the sql request
- For Oracle & SqlServer : execute a `select` query to check shema/table existence and execute  a `create` if the schema/table does not exist.
- If an exception is thrown, do not suppress it.

Reply:
The approach sounds good to me.

@brfrn169 (sorry for the interrupt) What do you think?

Reply:
The approach @Torch3333 mentioned is fine with me. 

Actually, the logic in the schema tool checks the JDBC implementation-specific message, which is the same as the current one. That is because I didn't want to have different logics (queries) between MySQL/PostgreSQL and Oracle/SQLServer, which I thought was a bit hacky. However, now that there are similar hacky logics in other parts, so we no longer have to worry about that anymore. 😅

Reply:
@brfrn169 Thank you! (Sorry for the interruption!)